Mirror subject.name.strategy from Java client

delphyne opened the pull request:
Provides a means to configure a subject name strategy for both key and value when using
the schema registry.  This brings the python client in line with the java client, and
enables the use of multiple schemas on a single topic.

delphyne commented:
[clabot:check]

rnpridgeon commented on setup.py:
```diff
@@ -45,4 +45,11 @@ def get_install_requirements(path):
           'avro': ['fastavro', 'requests', avro],
           'dev': get_install_requirements("test-requirements.txt")
+      },
+      entry_points={
```
As  mentioned above this should be the responsibility of the application as opposed to the `confluent-kafka-python` project.

delphyne commented on confluent_kafka/avro/__init__.py:
```diff
@@ -36,11 +37,28 @@ def __init__(self, config, default_key_schema=None,
         self._value_schema = default_value_schema
 
+    @staticmethod
+    def _load_name_strategy(config, key):
```
to be clear, is your recommendation to additional optional args to the ctor?  that's easy enough to get done; i only went this way because it matched the behavior of the class as it exists today.

delphyne commented:
@rnpridgeon are there any further changes you'd like to see on this PR?

nao921 commented:
Would be wonderful if this can be merged by someone with write access.

soxofaan commented:
@delphyne   is this PR still alive? 
Thanks for your work, we could really use this feature.

delphyne commented:
I believe I have addressed all feedback from @rnpridgeon, but he has not commented since July 14.  I am uncertain if he or @edenhill intend to accept this contribution or not.

delphyne commented on setup.py:
```diff
@@ -23,8 +23,12 @@
+
 def get_install_requirements(path):
-    content = open(os.path.join(os.path.dirname(__file__), path)).read()
+    content = open(os.path.join(install_requirements_path, path)).read()
```
it looks like this has since been fixed on master, so this change has now been removed.

delphyne commented on confluent_kafka/avro/__init__.py:
```diff
@@ -45,8 +46,11 @@ def __init__(self, config, default_key_schema=None,
+                                             key_subject_name_strategy=key_subject_name_strategy,
+                                             value_subject_name_strategy=value_subject_name_strategy)
+
         super(AvroProducer, self).__init__(config)
```
this has been addressed.

mhowlett commented:
Confluent Platform needs to allow for data with different schemas to be present in a single topic - people are moving past the simple data integration and buffering use cases of Kafka, and this feature enables an important design pattern (that we're evangelizing) that will become increasingly requested. Allowing people to do this is table-stakes for a streaming platform IMO. What we probably want to do instead is get associated changes made to connect. I haven't thought about the specifics in any depth at all - it's possible the @ept's changes are not the optimal way to go about this when viewed from a CP wide context, but would guess they probably are, and if not, well python is in the same boat as java ...

rnpridgeon commented:
Quick clarification: You can have multiple schemas in a single topic they just need to be compatible with one another. This is an important distinction as we talk about the scope of schema lineage enforcement. 

I see your point though @mhowlett  and given that this will likely be coming to .NET in the near future as well I will reevaluate my decision. I think there are are a few supporting features that may need to come into the picture as well in order to do this truly effective. I need to wrap up 1.0 and a few other things but after that I will put some careful thought into how to make this work given the state of the Avro Python implementation. From there we can reevaluate the existing PRs and make a plan on how to proceed.

 @delphyne sorry for falling behind on this, I will circle back with you after the release to see what we can do to move this forward.

soxofaan commented:
FWIW, I'd also like to chime in that this feature request is _not only_ about "multiple schemas in same topic".  I think there are also other use cases where one wants to go beyond the 1-on-1 coupling of schema and topic:
- same schema on multiple topics
- bake schema at compile time in your app and determine topic at run time

Now that we switched our app to the "record name strategy" things seem to make more sense and  the  "topic name strategy" feels like a weird default actually.

rnpridgeon commented:
This will be covered with generic serde support, #502
